### ISSUE TITLE
Allow for zero garbage JDBC monitoring

### ISSUE THREAD

**Issue by felixbarny** [p6spy/p6spy#436]
I'm now working on https://github.com/elastic/apm-agent-java where I have a focus on reducing the garbage the agent creates to a minimum.

It might be a bit out of scope for p6spy to implement an object pool which is used to recycle objects like `StatementWrapper`. But it would be nice if p6spy had hooks where you could add something like a custom `ObjectFactory` which abstracts all `new` invocations. P6spy could ship with a simple implementation which creates objects each time.

Other than creating objects, the `ObjectFactory` also has to be notified when the life cycle of an object ends (`close` is called), so that it can reset the state of the object and add it to the object pool again. For the built-in Implementation, this would be a noop. Also, all recyclable objects need to have a `reset` method which clears the state. 

A potential problem here is that this assumes, a `Statement` is not used anymore after it has been `close`d. If it does get used afterwards, that would lead to NPEs or, even worse, to calling a completely different statement.

Most of that can probably be implemented in a backwards compatible way. The `StatementWrapper` for example, is not directly instantiated with `new` but rather it's static `wrap` method gets called. https://github.com/p6spy/p6spy/blob/8be6672fd01f880065483be00a277baf6dad8479/src/main/java/com/p6spy/engine/wrapper/StatementWrapper.java#L47

What do you think about this proposal? If you like the idea, I can create a PR.

**Comment by typekpb** [comment]
Sounds like a good idea to me. Especially if we have a volunteer for
implementation.
Thank you!


On Mar 22, 2018 11:53, "Felix" <notifications@github.com> wrote:

I'm now working on https://github.com/elastic/apm-agent-java where I have a
focus on reducing the garbage the agent creates to a minimum.

It might be a bit out of scope for p6spy to implement an object pool which
is used to recycle objects like StatementWrapper. But it would be nice if
p6spy had hooks where you could add something like a custom ObjectFactory
which abstracts all new invocations. P6spy could ship with a simple
implementation which creates objects each time.

Other than creating objects, the ObjectFactory also has to be notified when
the life cycle of an object ends (close is called), so that it can reset
the state of the object and add it to the object pool again. For the
built-in Implementation, this would be a noop. Also, all recyclable objects
need to have a reset method which clears the state.

A potential problem here is that this assumes, a Statement is not used
anymore after it has been closed. If it does get used afterwards, that
would lead to NPEs or, even worse, to calling a completely different
statement.

Most of that can probably be implemented in a backwards compatible way. The
StatementWrapper for example, is not directly instantiated with new but
rather it's static wrap method gets called. https://github.com/p6spy/
p6spy/blob/8be6672fd01f880065483be00a277baf6dad8479/src/main/java/com/
p6spy/engine/wrapper/StatementWrapper.java#L47

What do you think about this proposal? If you like the idea, I can create a
PR.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/p6spy/p6spy/issues/436>, or mute the thread

**Comment by felixbarny** [comment]
Do you think it is safe to assume that there are no interactions with a Connection/Statement after it has been `close`d?

**Comment by typekpb** [comment]
Not sure, to be honest, maybe @quintonm has an idea.

**Comment by gavlyukovskiy** [comment]
@felixbarny at least `Connection#isValid` may be called after connection is closed. In fact I found few places in our code where connection was closed twice so I had to handle it in my listeners, but when I added test case I found that DBCP2 sets delegate to null after close and second interaction fails with NPE, so assuming they didn't provide a fix this is not the common case. 
HikariCP [sets delegate to closed connection instance](https://github.com/brettwooldridge/HikariCP/blob/dev/src/main/java/com/zaxxer/hikari/pool/ProxyConnection.java#L259) which can handle very few methods and throws `SQLException` on all the others.

On the other side p6spy collects all prepared statements' parameters so it won't be really "zero" garbage.
